### PR TITLE
[Pluggable storage] sync tasks and updaters refactor

### PR DESCRIPTION
**Pull request by EmilianoSanchez**
# Javascript commons library

## What did you accomplish?

- Separated updaters from sync task factories.

## How do we test the changes introduced in this PR?

- NA

## Extra Notes

### CODE REVIEW

**Review by mmelograno — APPROVED** [review]

File: src/sync/polling/updaters/segmentChangesUpdater.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+   * Segments updater returns a promise that resolves with a `false` boolean value if it fails at least to fetch a segment or synchronize it with the storage.
+   * Thus, a false result doesn't imply that SDK_SEGMENTS_ARRIVED was not emitted.
+   *
+   * @param {string[] | undefined} segmentNames list of segment names to fetch. By passing `undefined` it fetches the list of segments registered at the storage
+   * @param {boolean | undefined} noCache true to revalidate data to fetch on a SEGMENT_UPDATE notifications.
+   * @param {boolean | undefined} fetchOnlyNew if true, only fetch the segments that not exists, i.e., which `changeNumber` is equal to -1.
+   * This param is used by SplitUpdateWorker on server-side SDK, to fetch new registered segments on SPLIT_UPDATE notifications.
+   */
+  return function segmentChangesUpdater(segmentNames?: string[], noCache?: boolean, fetchOnlyNew?: boolean) {
+    log.debug(logPrefixSyncSegments + 'Started segments update');
```
Comment (emilianofant-split) [inline src/sync/polling/updaters/segmentChangesUpdater.ts:46]: Prefer `template string`
`log.debug(\`${logPrefixSyncSegments} Started segments update\`);`

File: src/sync/polling/updaters/segmentChangesUpdater.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    if (fetchOnlyNew) segments = segments.filter(segmentName => segmentsCache.getChangeNumber(segmentName) === -1);
+
+    // Async fetchers are collected here.
+    const updaters: Promise<number>[] = [];
+
+    for (let index = 0; index < segments.length; index++) {
+      const segmentName = segments[index];
+      const since = segmentsCache.getChangeNumber(segmentName);
+
+      log.debug(logPrefixSyncSegments + `Processing segment ${segmentName}`);
```
Comment (emilianofant-split) [inline src/sync/polling/updaters/segmentChangesUpdater.ts:59]: Prefer `template string`
``` log.debug(`${logPrefixSyncSegments} Processing segment ${segmentName}`);```

File: src/sync/polling/updaters/segmentChangesUpdater.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        let changeNumber = -1;
+        changes.forEach(x => {
+          if (x.added.length > 0) segmentsCache.addToSegment(segmentName, x.added);
+          if (x.removed.length > 0) segmentsCache.removeFromSegment(segmentName, x.removed);
+          if (x.added.length > 0 || x.removed.length > 0) {
+            segmentsCache.setChangeNumber(segmentName, x.till);
+            changeNumber = x.till;
+          }
+
+          log.debug(logPrefixSyncSegments + `Processed ${segmentName} with till = ${x.till}. Added: ${x.added.length}. Removed: ${x.removed.length}`);
```
Comment (emilianofant-split) [inline src/sync/polling/updaters/segmentChangesUpdater.ts:71]: Prefer `template string`

File: src/sync/polling/updaters/segmentChangesUpdater.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        });
+
+        return changeNumber;
+      }));
+
+    }
+
+    return Promise.all(updaters).then(shouldUpdateFlags => {
+      // if at least one segment fetch successes, mark segments ready
```
Comment (emilianofant-split) [inline src/sync/polling/updaters/segmentChangesUpdater.ts:80]: typo: 
`if at least one segment fetch succeeded , mark segments ready`

File: src/sync/polling/updaters/splitChangesUpdater.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+function checkAllSegmentsExist(segmentsStorage: ISegmentsCacheSync) {
+  return segmentsStorage.getRegisteredSegments().every(segmentName => segmentsStorage.getChangeNumber(segmentName) !== -1);
+}
+
+/**
+ * Collect segments from a raw split definition.
+ * Exported for testing purposes.
+ */
+export function parseSegments({ conditions }: ISplit): ISet<string> {
+
```
Comment (emilianofant-split) [inline src/sync/polling/updaters/splitChangesUpdater.ts:26]: Remove this empty line.

**Review by emilianofant-split — APPROVED** [review]
LGTM

File: src/sync/polling/updaters/segmentChangesUpdater.ts (hunk 1 above)
Comment (EmilianoSanchez) [reply src/sync/polling/updaters/segmentChangesUpdater.ts:46]: done

**Pull request merged**